### PR TITLE
Installation of cats-modules

### PR DESCRIPTION
**Pull request by BabichMikhail**

### CODE REVIEW

**Comment by sdogerg** [comment]
g

File: install.pl [hunk 1]
```diff
@@ -153,6 +153,11 @@ sub get_modulesdir {
     $modulesdir;
 }
 
+step 'Creating judge_next_id.txt file', sub {
+    open my $fnext_id, '>', 'judge_next_id.txt' or die "Can't open judge_next_id.txt file";
+    printf $fnext_id "%d\n", 100;
+};
+
```
Comment (klenin) [inline install.pl]: That is totally unneeded. Set default in `cats-problem/new_id` instead.

File: .gitmodules [hunk 2]
```diff
@@ -4,3 +4,6 @@
 [submodule "spawner-bin"]
 	path = spawner-bin
 	url = https://github.com/klenin/spawner-bin.git
+[submodule "lib/cats-modules"]
+	path = lib/cats-modules
+	url = https://github.com/BabichMikhail/cats-modules
```
Comment (klenin) [inline .gitmodules]: Let's move repository before the merge.
See Settings->Transfer ownership in GitHub UI.

File: install.pl [hunk 3]
```diff
@@ -139,3 +139,46 @@ sub step_copy {
...
+    $xml_parser->parsefile('config.xml');
+    $modulesdir;
+}
+
+step 'Installing cats-modules', sub {
+    -e 'judge_next_id.txt' or die 'Missing file judge_next_id.txt';
+    require XML::Parser::Expat;
+    my $modulesdir = get_modulesdir();
+    my $cats_modules_dir = File::Spec->catfile(qw[lib cats-modules]);
+    open my $modules_list, '<', File::Spec->catfile($cats_modules_dir, 'modules.txt') or die "Can't open modules_list.txt";
```
Comment (klenin) [inline install.pl]: Wrong file name in error message.

File: install.pl [hunk 4]
```diff
@@ -139,3 +139,46 @@ sub step_copy {
...
+        my ($p, $el, %atts) = @_;
+        $el eq 'judge' or return;
+        $modulesdir = $atts{modulesdir};
+    });
+    $xml_parser->parsefile('config.xml');
+    $modulesdir;
+}
+
+step 'Installing cats-modules', sub {
+    -e 'judge_next_id.txt' or die 'Missing file judge_next_id.txt';
```
Comment (klenin) [inline install.pl]: Remove this.

File: install.pl [hunk 5]
```diff
@@ -139,3 +139,46 @@ sub step_copy {
...
+    require XML::Parser::Expat;
+    my $modulesdir = get_modulesdir();
+    my $cats_modules_dir = File::Spec->catfile(qw[lib cats-modules]);
+    open my $modules_list, '<', File::Spec->catfile($cats_modules_dir, 'modules.txt') or die "Can't open modules_list.txt";
+    my (%path, @list);
```
Comment (klenin) [inline install.pl]: `@list` is a bad name

File: install.pl [hunk 6]
```diff
@@ -139,3 +139,46 @@ sub step_copy {
...
+    for (<$modules_list>) {
+        $_ =~ s/\n//;
+        my $module = File::Spec->catfile($cats_modules_dir, $_);
+        $path{$module} = globq(File::Spec->catfile($module, '*.xml'));
+        push @list, $module;
+    }
+    print "\n";
+    for (@list) {
+        !system("local_judge.cmd --problem $_ >>nul 2>&1") or next;
```
Comment (klenin) [inline install.pl]: `>nul 2>nul`

File: judge.pl [hunk 7]
```diff
@@ -949,7 +949,7 @@ sub process_request
     }
     $judge->save_log_dump($r, $log->{dump});
     $judge->set_request_state($r, $state, %$r);
-    return if $state == $cats::st_unhandled_error;
+    return if !$judge->{solution} || $state == $cats::st_unhandled_error;
```
Comment (klenin) [inline judge.pl]: This commit should come before the 'Add installation of cats-modules'

**Pull request merged**